gusto/state.py

```diff
@@ -278,15 +278,47 @@ def setup_diagnostics(self):
         """
         for name in self.output.perturbation_fields:
             f = Perturbation(name)
-            self.diagnostic_fields.insert(0, f)
+            self.diagnostic_fields.append(f)
 
         for name in self.output.steady_state_error_fields:
             f = SteadyStateError(self, name)
-            self.diagnostic_fields.insert(0, f)
-
-        for diagnostic in self.diagnostic_fields:
-            diagnostic.setup(self)
-            self.diagnostics.register(diagnostic.name)
+            self.diagnostic_fields.append(f)
+
+        # this list will contain the diagnostic fields in the right order for
+        # calculation
+        reordered_diagnostic_fields = []
+        # this list contains the diagnostic fields that have not yet
+        # been set up (currently all of them)
+        remaining_diagnostic_fields = self.diagnostic_fields.copy()
+        repeat = True
+        while repeat:
+            # this list contains the diagnostic fields that cannot be
+            # set up on this sweep due to dependence on other field
+            # that have not yet been set up
+            postponed_fields = []
+            for diagnostic in remaining_diagnostic_fields:
+                try:
+                    # try to setup the diagnostic and append to reordered list
+                    diagnostic.setup(self)
+                    self.diagnostics.register(diagnostic.name)
+                    reordered_diagnostic_fields.append(diagnostic)
+                except RuntimeError:
+                    # if the setup gives a RuntimeError then postpone this field
+                    postponed_fields.append(diagnostic)
+                    logger.info("setup of diagnostic %s postponed" % diagnostic.name)
+            if len(postponed_fields) == 0:
+                # if nothing needs postponing then we're done
+                repeat = False
+            elif len(postponed_fields) < len(remaining_diagnostic_fields):
+                # if we've managed to reduce the number of fields we
+                # postpone then continue with another sweep
+                remaining_diagnostic_fields = postponed_fields.copy()
+            else:
+                # if we're not managing to reduce the number of fields
+                # we postpone then something bad is happening
+                raise RuntimeError("There is a problem with your diagnostic fields")
+        # update the list of diagnostic fields to the reordered version
+        self.diagnostic_fields = reordered_diagnostic_fields
```

Comment:
You want a topological sort of the field dependency graph, which should be a DAG:

Here, I've implemented [Kahn's algorithm](https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm).
```python
def topo_sort(field_deps):
    # map node: (input_deps, output_deps)
    graph = dict((f, (list(deps), [])) for f, deps in field_deps)
    roots = []
    for f, input_deps in field_deps:
        if len(input_deps) == 0:
            # No dependencies, candidate for evaluation
            roots.append(f)
        for d in input_deps:
            # add f as output dependency
            graph[d][1].append(f)

    schedule = []
    while roots:
        n = roots.pop()
        schedule.append(n)
        output_deps = list(graph[n][1])
        for m in output_deps:
            # Remove edge
            graph[m][0].remove(n)
            graph[n][1].remove(m)
            # If m now as no input deps, candidate for evaluation
            if len(graph[m][0]) == 0:
                roots.append(m)
    if any(len(i) for i, _ in graph.values()):
        cycle = "\n".join("%s -> %s" % (f, i) for f, (i, _) in graph.items()
                          if f not in schedule)
        raise RuntimeError("Field dependencies have a cycle:\n\n%s" % cycle)
    return schedule
```

Run as:

```python
field_deps = [("a", ("b", "c"),),
              ("b", ("c", "d"),),
              ("c", ()),
              ("d", ("c", ))]

schedule = topo_sort(field_deps)
```

If there is a cycle, you get a runtime error:

```python
field_deps = [("a", ("b", "c"),),
              ("b", ("c", "d"),),
              ("c", ()),
              ("d", ("c", "b"))]
topo_sort(field_deps)

=> RuntimeError: Field dependencies have a cycle:

a -> ['b']
b -> ['d']
d -> ['b']
```